src/main.rs

```diff
@@ -2542,24 +2562,55 @@ fn run_steps(ib: &mut ImageBuilder) -> Result<()> {
 
                 #[derive(Deserialize)]
                 struct DirArgs {
+                    // The name and metadata of the directrory to be created.
                     dir: String,
                     owner: String,
                     group: String,
                     mode: String,
+
+                    // An optional source file of the directory, if contents are
+                    // to be copied, rather than simply creating the directory.
+                    extsrc: Option<String>
                 }
 
                 let a: DirArgs = step.args()?;
                 let owner = translate_uid(&a.owner)?;
                 let group = translate_gid(&a.group)?;
+                let extsrc = ib.expando(a.extsrc.as_deref())?;
 
                 if !a.dir.starts_with('/') {
                     bail!("dir must be fully qualified path");
                 }
-                let dir = format!("{}{}", targmp, a.dir);
+                let dst = format!("{}{}", targmp, a.dir);
 
                 let mode = u32::from_str_radix(&a.mode, 8)?;
 
-                ensure::directory(log, &dir, owner, group, mode)?;
+                ensure::directory(log, &dst, owner, group, mode)?;
+
+                if let Some(extsrc) = &extsrc {
+                    if extsrc.starts_with('/') {
+                        bail!("external source directory must be a relative path");
+                    }
+                    let dst = PathBuf::from(dst);
+
+                    // "src" is the absolute path to a directory to be copied.
+                    let src = ib.external_src_dir(extsrc)?;
+
+                    for entry in walkdir::WalkDir::new(&src) {
+                        let entry = entry?;
+                        let from = entry.path();
+                        let to = dst.join(from.strip_prefix(&src)?);
+                        if entry.file_type().is_dir() {
+                            if let Err(e) = std::fs::create_dir(to) {
+                                if !matches!(e.kind(), std::io::ErrorKind::AlreadyExists) {
+                                    return Err(e.into());
+                                }
+                            }
+                        } else if entry.file_type().is_file() {
+                            std::fs::copy(from, to)?;
+                        }
+                    }
+                }
```

Comment:
This is the meat of this PR - I more-or-less mimicked how `ensure_file` seems to work with pulling `extsrc` files, but expanded that to work for directories too.